src/NHibernate/Loader/Loader.cs

```diff
@@ -981,9 +981,18 @@ private void CheckVersion(int i, IEntityPersister persister, object id, object e
 						obj =
 							InstanceNotYetLoaded(rs, i, persister, key, lockModes[i], optionalObjectKey,
 												 optionalObject, hydratedObjects, session);
+
+						// IUniqueKeyLoadable.CacheByUniqueKeys caches all unique keys of the entity, regardless of
+						// associations loaded by the query. So if the entity is already loaded, it has forcibly already
+						// been cached too for all its unique keys, provided its persister implement it. With this new
+						// way of caching unique keys, it is no more needed to handle caching for alreadyLoaded path
+						// too.
+						(persister as IUniqueKeyLoadable).CacheByUniqueKeys(obj, session);
```

Comment:
Here is the change for avoiding cache misses when looking-up by unique key.

Due to #1226, this caching of unique key has a bit evolved.
Prior to #1226, the caching by unique key was done only for entities not yet loaded, when they were loaded through a query with eager loads referencing the entity by its unique key and as their RHS, excepted in the many-to-many case (which was failing and is still failing).
For fixing #1226, this caching has been moved for occurring also with entities already loaded.
Now with this PR, it is moved back to the path of entities not yet loaded, and no more done for this already loaded. This is done because now, instead of caching only by unique key used by RHS side of associations loaded by the query, it is done for all alternate unique keys declared in the entity. So it does no more depends on which associations are eagerly  loaded in the query loading the entity, the cache feeding will be done wherever the entity is actually loaded.

This is an overhead for session usages which were not doing any eager load, but I think it is acceptable, because anyway the unique key caching is actually done only in case of `property-ref` usage in mappings, feature which should be used as little as possible. (Mapped unique keys do not trigger this caching. Only the property target of a property-ref is cached as a unique key.)